### PR TITLE
Fix secure/noise by securing explicitly

### PR DESCRIPTION
**Pull request by sinkingsugar**
Also added noise to pubsub tests

### CODE REVIEW

File: libp2p/standard_setup.nim [hunk 1]
```diff
@@ -1,11 +1,20 @@
+# compile time options here
+const
+  libp2p_secure {.strdefine.} = ""
```
Comment (arnetheduck) [inline libp2p/standard_setup.nim:3]: can this be negotiated / dynamic?

Comment (sinkingsugar) [reply libp2p/standard_setup.nim:3]: Well, like @dryajov said in the call yesterday:
1. multistream select as it stands has no ordering warranty
2. for NBC noise should be the only one allowed
So we decided to use compile time flags.

Indeed for other generic usages it should be just added to the table in the same file to be an option of the 2 between secio and noise.

File: libp2p/protocols/secure/secure.nim [hunk 2]
```diff
@@ -29,7 +29,7 @@ method writeMessage*(c: SecureConn, data: seq[byte]) {.async, base.} =
 
 method handshake(s: Secure,
                  conn: Connection,
-                 initiator: bool = false): Future[SecureConn] {.async, base.} =
+                 initiator: bool): Future[SecureConn] {.async, base.} =
```
Comment (dryajov) [inline libp2p/protocols/secure/secure.nim:32]: why remove the default?

File: libp2p/protocols/secure/secure.nim [hunk 3]
```diff
@@ -54,7 +54,7 @@ proc readLoop(sconn: SecureConn, conn: Connection) {.async.} =
       await sconn.close()
     trace "ending Secure readLoop"
 
-proc handleConn*(s: Secure, conn: Connection, initiator: bool = false): Future[Connection] {.async, gcsafe.} =
+proc handleConn*(s: Secure, conn: Connection, initiator: bool): Future[Connection] {.async, gcsafe.} =
```
Comment (dryajov) [inline libp2p/protocols/secure/secure.nim:57]: why remove the default?

File: libp2p/protocols/secure/secure.nim [hunk 4]
```diff
@@ -80,9 +80,9 @@ method init*(s: Secure) {.gcsafe.} =
 
   s.handler = handle
 
-method secure*(s: Secure, conn: Connection): Future[Connection] {.async, base, gcsafe.} =
+method secure*(s: Secure, conn: Connection, initiator: bool): Future[Connection] {.async, base, gcsafe.} =
   try:
-    result = await s.handleConn(conn, true)
+    result = await s.handleConn(conn, initiator)
```
Comment (dryajov) [inline libp2p/protocols/secure/secure.nim:85]: `secure` is explicitly an initiator, we should leave it like that.

File: libp2p/standard_setup.nim (hunk 1 above)
Comment (dryajov) [reply libp2p/standard_setup.nim:3]: @arnetheduck it is negotiated dynamically but the order of why one is chosen isn't guaranteed, so we want to always run with either secio or noise for now.

File: libp2p/switch.nim [hunk 5]
```diff
@@ -191,7 +191,7 @@ proc upgradeIncoming(s: Switch, conn: Connection) {.async, gcsafe.} =
                        {.async, gcsafe, closure.} =
     trace "Securing connection"
     let secure = s.secureManagers[proto]
-    let sconn = await secure.secure(conn)
+    let sconn = await secure.secure(conn, false)
```
Comment (dryajov) [inline libp2p/switch.nim:194]: we should keep using the default here.

File: libp2p/protocols/secure/secure.nim (hunk 2 above)
Comment (arnetheduck) [reply libp2p/protocols/secure/secure.nim:32]: why have a default? for every intiiator, isn't there a non-initiator? why is one more important than the opther?

Comment (sinkingsugar) [reply libp2p/protocols/secure/secure.nim:32]: simply cos defaults are a bad habit and dangerous generally, specially here had no use anymore

File: libp2p/protocols/secure/secure.nim (hunk 4 above)
Comment (sinkingsugar) [reply libp2p/protocols/secure/secure.nim:85]: That was the whole issue with this PR, sadly it's not, check `switch.nim` implementation

File: libp2p/switch.nim (hunk 5 above)
Comment (sinkingsugar) [reply libp2p/switch.nim:194]: Any reason why being explicit is bad here tho?
Helps anyone reading the code imho

File: libp2p/protocols/secure/secure.nim (hunk 2 above)
Comment (dryajov) [reply libp2p/protocols/secure/secure.nim:32]: > why have a default? for every intiiator, isn't there a non-initiator? why is one more important than the opther?

Historically, `secure` was called to initiate the flow and there was no flag to indicate who is the initiator because `secure` already implied it. The distinction came with Noise, the default was there to not break/change the public API. 

Why does Noise have this requirement and are we signaling it properly in all the right places?

File: libp2p/switch.nim (hunk 5 above)
Comment (dryajov) [reply libp2p/switch.nim:194]: Nothing against it, I'm just concerned we're changing the intent of the API and there is no reasoning for it anywhere. Granted that the flag was introduced in prior commits, but I can't find anything that explains the change properly.

File: libp2p/protocols/secure/secure.nim (hunk 2 above)
Comment (dryajov) [reply libp2p/protocols/secure/secure.nim:32]: The reasoning behind the default is that not all securing mechanisms need this flag, for example, secio doesn't use it.

> simply cos defaults are a bad habit and dangerous generally, specially here had no use anymore

This is a bit of a generalization. Every tool has it's use and not everything is a nail that needs a hammer - https://en.wikipedia.org/wiki/Law_of_the_instrument

Comment (sinkingsugar) [reply libp2p/protocols/secure/secure.nim:32]: The reason for Noise is handshake patterns are different between initiator/receiver:
https://noiseprotocol.org/noise.html#handshake-patterns

**Comment by sinkingsugar** [comment]
> LGTM, but lets keep the default for the `initiator` flag please, unless there is a very good reason not to which I can't see right now.

If the reason around the change in switch is clear, should I restore default value and merge ?

File: libp2p/protocols/secure/secure.nim (hunk 2 above)
Comment (arnetheduck) [reply libp2p/protocols/secure/secure.nim:32]: > The reasoning behind the default is that not all securing mechanisms need this flag, for example, secio doesn't use it.

This is an argument to not have a default, right? I mean, if I'm coming from `secio`, I would want to know about this important difference between the two protocols, or my application won't work and I'll have to spend time looking for the cause - finding a default boolean is hard.

Basically, there's no meaningful way to use the protocol with _only_ the default setting.

**Comment by dryajov** [comment]
> > LGTM, but lets keep the default for the `initiator` flag please, unless there is a very good reason not to which I can't see right now.
> 
> If the reason around the change in switch is clear, should I restore default value and merge ?

All good, the change makes sense, the default isn't suitable anymore.
